### PR TITLE
SyncBatchNorm Update on input dimension checks

### PR DESCRIPTION
**Pull request by jjsjann123**
update the requirements on input dimensions for `torch.nn.SyncBatchNorm`:
1. 2D inputs is now permissible, #20204 ;
2. requires at least two element along normalization plane (BatchNorm behavior);

### CODE REVIEW

File: torch/nn/modules/_functions.py [hunk 1]
```diff
@@ -7,7 +7,11 @@ class SyncBatchNorm(Function):
     @staticmethod
     def forward(self, input, weight, bias, running_mean, running_var, eps, momentum, process_group, world_size):
         input = input.contiguous()
-        count = torch.Tensor([input.numel() // input.size(1)]).to(input.device)
+
+        size = input.numel() // input.size(1)
+        if size == 1:
+            raise ValueError('Expected more than 1 value per channel when training, got input size {}'.format(size))
```
Comment (albanD) [inline torch/nn/modules/_functions.py]: Isn't this method called both for training and testing?

Comment (jjsjann123) [reply torch/nn/modules/_functions.py]: Good point, missed the training flag: https://github.com/pytorch/pytorch/blob/5a8ad66354c81f1fbe1e68848b420cad7a2ce97b/torch/nn/functional.py#L1651-L1666

Comment (jjsjann123) [reply torch/nn/modules/_functions.py]: Forgot how to English.

The default behavior is to only raise exception for `self.training or not self.track_running_stats`, which is the same logic that would take you to `SyncBatchNorm`.
So we should be good here.

Comment (jjsjann123) [reply torch/nn/modules/_functions.py]: The missing link here: https://github.com/pytorch/pytorch/blob/5a8ad66354c81f1fbe1e68848b420cad7a2ce97b/torch/nn/modules/batchnorm.py#L78

**Comment by albanD** [comment]
The CI failures look legitimate. Could you fix them please?

File: torch/nn/modules/_functions.py (hunk 1 above)
Comment (albanD) [reply torch/nn/modules/_functions.py]: Thanks for the links !

**Comment by jjsjann123** [comment]
Sorry for the delayed response, only got a machine earlier today to repro the error.

Looks like numerical issues, the parameters only show divergence at iteration 5.
```
OrderedDict([('weight', tensor([33.3129, 33.9196], device='cuda:2')), ('bias', tensor([ 7.4568, -0.9556], device='cuda:2')), ('running_mean', tensor([-0.1032,  0.2549], device='cuda:2')), ('running_var', tensor([0.6850, 0.8714], device='cuda:2')), ('num_batches_tracked', tensor(5, device='cuda:2'))])

OrderedDict([('module.weight', tensor([33.3128, 33.9196], device='cuda:2')), ('module.bias', tensor([ 7.4568, -0.9556], device='cuda:2')), ('module.running_mean', tensor([-0.1032,  0.2549], device='cuda:2')), ('module.running_var', tensor([0.6850, 0.8714], device='cuda:2')), ('module.num_batches_tracked', tensor(5, device='cuda:2'))])
```
It actually doesn't have anything to do with the SyncBatchNorm specifically. If I disable cudnn and run the base forward with native batch_norm, the error goes away.

I'll update the test to avoid the false alarm.

**Comment by albanD** [comment]
I am not very familiar with this particular test. 
Why is it tripped by cudnn's implementation in your case but not in the others? Is it just by chance in the 2D case?

**Comment by jjsjann123** [comment]
In other tests here, It's the Softmax() they have. which reduces the error & gradient propagated back, that covers up the numerical issue.
Personally, I think it makes more sense to fallback to native BN, which rules out the numerical issue other than magically hide it with the model. But I don't have a strong favor here and I can also change my model to append a softmax in the end. Please let me know which one you prefer.

**Comment by albanD** [comment]
I'm happy with your version.
Could you add a small comment next to the context manager that disables cudnn to keep track of that.
It will be good to merge afterwards.

**Review by albanD — APPROVED** [review]
Thanks !

**Comment by jjsjann123** [comment]
Oops, forgot to change the flag back after I verified the softmax fix. :man_facepalming:

**Comment by albanD** [comment]
Could you rebase on top of master please because it seems like there is some unrelated consistent fails in the CI here.

**Comment by jjsjann123** [comment]
Failed test doesn't seem to be related, should I try merge master again?

**Comment by albanD** [comment]
Looks fine, I'll try to merge.
